swabbie-redis/src/main/kotlin/com/netflix/spinnaker/swabbie/redis/RedisResourceTrackingRepository.kt

```diff
@@ -34,76 +35,107 @@ class RedisResourceTrackingRepository(
+  }
+
+  override fun getIdsOfMarkedResourcesToDelete(): Set<String> {
+    return getAllIds(DELETE_KEY, false)
+  }
+
+  private fun getAllIds(key: String, includeAll: Boolean): Set<String> {
+    return redisClientDelegate.run {
+      this.withCommandsClient<Set<String>> { client ->
+        if (includeAll) {
+          client.zrangeByScore(key, "-inf", "+inf")
+        } else {
+          client.zrangeByScore(key, 0.0, clock.instant().toEpochMilli().toDouble())
+        }
       }
     }
+  }
 
-  override fun upsert(markedResource: MarkedResource, score: Long) {
-    "${markedResource.namespace}:${markedResource.resourceId}".let { id ->
-      markedResource.apply {
-        markTs = if (markTs != null) markTs else Instant.now(clock).toEpochMilli()
-        updateTs = if (markTs != null) Instant.now(clock).toEpochMilli() else null
+  private fun hydrateMarkedResources(resourseIds: Set<String>): List<MarkedResource> {
+    if (resourseIds.isEmpty()) return emptyList()
+    return redisClientDelegate.run {
+      this.withCommandsClient<Set<String>> { client ->
+        client.hmget(SINGLE_RESOURCES_KEY, *resourseIds.map { it }.toTypedArray()).toSet()
+      }.map { json ->
+        objectMapper.readValue<MarkedResource>(json)
       }
+    }
+  }
 
-      resourceKey(id).let { key ->
-        redisClientDelegate.withCommandsClient { client ->
-          client.hset(SINGLE_RESOURCES_KEY, id, objectMapper.writeValueAsString(markedResource))
-          client.zadd(ALL_RESOURCES_KEY, score.toDouble(), id)
-        }
-      }
+  override fun upsert(markedResource: MarkedResource, deleteScore: Long, softDeleteScore: Long) {
+    val id = markedResource.uniqueId()
+
+    markedResource.apply {
+      markTs = if (markTs != null) markTs else Instant.now(clock).toEpochMilli()
+      updateTs = if (markTs != null) Instant.now(clock).toEpochMilli() else null
+    }
+
+    redisClientDelegate.withCommandsClient { client ->
+      client.hset(SINGLE_RESOURCES_KEY, id, objectMapper.writeValueAsString(markedResource))
+      client.zadd(DELETE_KEY, deleteScore.toDouble(), id)
+      client.zadd(SOFT_DELETE_KEY, softDeleteScore.toDouble(), id)
+    }
+  }
+
+  override fun setSoftDeleted(markedResource: MarkedResource) {
+    val id = markedResource.uniqueId()
+    redisClientDelegate.withCommandsClient { client ->
+      client.zrem(SOFT_DELETE_KEY, id)
     }
   }
 
   override fun remove(markedResource: MarkedResource) {
-    "${markedResource.namespace}:${markedResource.resourceId}".let { id ->
-      resourceKey(id).let { key ->
-        redisClientDelegate.withCommandsClient { client ->
-          client.zrem(ALL_RESOURCES_KEY, id)
-          client.hdel(SINGLE_RESOURCES_KEY, id)
-        }
-      }
+    val id = markedResource.uniqueId()
+    redisClientDelegate.withCommandsClient { client ->
+      client.zrem(DELETE_KEY, id)
+      client.zrem(SOFT_DELETE_KEY, id)
+      client.hdel(SINGLE_RESOURCES_KEY, id)
     }
```

Comment:
Would recommend adding retries where possible around redis operations. This isn't done in many other places in Spinnaker, but for the new persistence backends I've been working on, I've been ensuring retries are put around all interactions so transient network failures don't cause as many application errors.

Reply:
@robzienert happy to. Do you recommend using RetrySupport?

Reply:
Ok. Will fix in another PR.